crowbar_framework/app/models/service_object.rb

```diff
@@ -855,6 +855,36 @@ def violates_admin_constraint?(elements, role, nodes_is_admin = {})
     false
   end
 
+  def violates_platform_constraint?(elements, role)
+    if role_constraints[role] && role_constraints[role].has_key?("platform")
+      elements[role].each do |element|
+        next if is_cluster? element
+        node = NodeObject.find_node_by_name(element)
+
+        constraints = role_constraints[role]["platform"]
+        return !constraints.any? do |platform, version|
+          equal_or_match(node[:platform], platform) && equal_or_match(node[:platform_version], version)
+        end
+      end
+    end
+    false
+  end
+
+  def violates_exclude_platform_constraint?(elements, role)
+    if role_constraints[role] && role_constraints[role].has_key?("exclude_platform")
+      elements[role].each do |element|
+        next if is_cluster? element
+        node = NodeObject.find_node_by_name(element)
+
+        constraints = role_constraints[role]["exclude_platform"]
+        return constraints.any? do |platform, version|
+          equal_or_match(node[:platform], platform) && equal_or_match(node[:platform_version], version)
+        end
```

Comment:
There's the question of sharing the core of these two methods in a small helper. For instance:

```ruby
def matches_platform_constraint?(elements, constraints)
  elements.each do |element|
    next if is_cluster? element
    node = NodeObject.find_node_by_name(element)

    return !constraints.any? do |platform, version|
      equal_or_match(node[:platform], platform) && equal_or_match(node[:platform_version], version)
    end
  end
end

def violates_platform_constraint?(elements, role)
  if role_constraints[role] && role_constraints[role].has_key?("platform")
    matches_platform_constraint?(elements[role], role_constraints[role]["platform"])
  else
    false
  end
end